### PR TITLE
refactor: update to facet DirectGovernanceFarmer

### LINKED ISSUE #773: Move `DirectGovernanceFarmer` to diamond facets
We've migrated most of our contracts to the diamond proxy implementation. But not all of them.

What should be done:
- refactor [DirectGovernanceFarmer](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/DirectGovernanceFarmer.sol) to be part of the diamond [facets](https://github.com/ubiquity/ubiquity-dollar/tree/development/packages/contracts/src/dollar/facets) 
- add `DirectGovernanceFarmer` to the [deployment script](https://github.com/ubiquity/ubiquity-dollar/tree/development/packages/contracts/scripts/deploy/dollar/solidityScripting)
- refactor `DirectGovernanceFarmer` [tests](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/test/dollar/DirectGovernanceFarmer.t.sol) accordingly

### PR DESCRIPTION
**Pull request by molecula451**
Resolves #773

- [x] depositSingle
- [x] depositMulti
- [x] withdrawWithId
- [x] withdraw

Reimplemented original withdraw function with same named, diamond cut cannot add same name functions
Implemented initialize function mimicking a constructor towards starting the facet and its required values

### CODE REVIEW

**Comment by molecula451** [comment]
just look how big the test file is 😛 https://github.com/ubiquity/ubiquity-dollar/blob/abbb58606310f92893beacc02cc694bed25e3398/packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol @pavlovcik

File: packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol [hunk 1]
```diff
@@ -43,22 +43,33 @@ contract DirectGovernanceFarmerFacet is Modifiers {
...
         address operator,
         address from,
         uint256 id,
         uint256 value,
         bytes calldata data
-    ) external view returns (bytes4) {
+    ) external pure returns (bytes4) {
         return
-            LibDirectGovernanceFarmer.onERC1155Received(
+            LibDirectGovernanceFarmer.onERC1155ReceivedDirectGovernance(
```
Comment (molecula451) [inline packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: @rndquu renamed cause libcreditnft has one and diamond it's detecting naming clash

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol [hunk 2]
```diff
@@ -104,32 +104,25 @@ library LibDirectGovernanceFarmer {
      * called at the end of a `safeTransferFrom` after the balance has been updated.
      *
      * NOTE: To accept the transfer, this must return
-     * `bytes4(keccak256("onERC1155Received(address,address,uint256,uint256,bytes)"))`
-     * (i.e. 0xf23a6e61, or its own function selector).
+     * `bytes4(keccak256("onERC1155ReceivedDirectGovernance(address,address,uint256,uint256,bytes)"))`
+     * (i.e. 0x7abeba72, or its own function selector).
```
Comment (molecula451) [inline packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol]: @rndquu selector changed so

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 3]
```diff
@@ -108,19 +108,26 @@ contract DirectGovernanceFarmerFacetTest is DiamondTestSetup {
...
+        assertEq(
+            directGovernanceFarmerFacet.onERC1155ReceivedDirectGovernance(
+                address(stakingShareAddress),
+                address(stakingShareAddress),
+                3,
+                4,
+                ""
+            ),
+            directGovernanceFarmerFacet
+                .onERC1155ReceivedDirectGovernance
```
Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: but the test clears

**Comment by molecula451** [comment]
gentlemen finally on of the biggest contracts on the repo, it's ready to move to diamond, don't get me wrong the contract looks shrinked because of being facet, much smaller, but on the inside it's so big that the debugging was a cave, anyway, we are pushing and moving forward with this one

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol [hunk 4]
```diff
@@ -145,45 +142,46 @@ contract DirectGovernanceFarmer is ReentrancyGuard {
...
+        IERC20(token).safeIncreaseAllowance(
+            data.depositZapUbiquityDollar,
+            amount
         );
+        uint256 lpAmount = IDepositZap(data.depositZapUbiquityDollar)
+            .add_liquidity(data.ubiquity3PoolLP, tokenAmounts, 0);
 
         //STEP2: stake UAD3CRVf to Staking
+
         //TODO approve token to be transferred to Staking contract
```
Comment (gitcoindev) [inline packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:157]: since approve was implemented TODO comment can be deleted

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol [hunk 5]
```diff
@@ -340,34 +342,37 @@ contract DirectGovernanceFarmer is ReentrancyGuard {
...
+            .remove_liquidity(
+                data.ubiquity3PoolLP,
+                lpTokenAmount,
+                [uint256(0), uint256(0), uint256(0), uint256(0)]
+            ); //[Ubiquity Dollar, DAI, USDC, USDT]
+
+        IERC20(data.ubiquityDollar).safeTransfer(msg.sender, tokenAmounts[0]);
+        IERC20(data.token0).safeTransfer(msg.sender, tokenAmounts[1]);
+        IERC20(data.token1).safeTransfer(msg.sender, tokenAmounts[2]);
+        IERC20(data.token2).safeTransfer(msg.sender, tokenAmounts[3]);
```
Comment (gitcoindev) [inline packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:349]: It is good this is covered by tests, took me a while to go up and down and compare indexes, especially that indexes differ, e.g. token0 is in tokenAmounts[1]. Perhaps if token0, token1, token2 were renamed to token_DAI, token_USDC, token_USDT it would be more readable and less error prone. But I understand you move to facet so the original names were not changed.

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 6]
```diff
@@ -92,70 +91,180 @@ contract DirectGovernanceFarmerTest is LocalTestHelper {
...
+            1
+        );
+    }
+
+    function testDepositMultipleTokens_ShouldRevert_IfDurationIsNotValid()
+        public
+    {
         address userAddress = address(0x100);
         vm.prank(userAddress);
         vm.expectRevert("duration weeks must be between 1 and 208");
```
Comment (gitcoindev) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:200]: We could also add a boundary value check: 209 weeks

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 7]
```diff
@@ -257,27 +348,28 @@ contract DirectGovernanceFarmerTest is LocalTestHelper {
         vm.stopPrank();
 
         // mint 100 Dollars to user
-        vm.prank(admin);
-        dollar.mint(userAddress, 100e18);
+        vm.startPrank(admin);
+        dollar.mint(userAddress, 99e18);
```
Comment (gitcoindev) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: Why the comment says mint 100 Dollars and 99 Dollars are minted?

**Review by gitcoindev — CHANGES_REQUESTED** [review]
In general a good job, I left a few comments. After they are addressed I will approve from my side and pass on to @rndquu for the review.

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol (hunk 4 above)
Comment (molecula451) [reply packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:157]: no, no deleting, this will guideful on a future PR, but not on this, we need this to get in, then we check it out this one

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol (hunk 5 above)
Comment (molecula451) [reply packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:349]: absolutely, one of the purposes is that im keeping on original as possible with minimal changes, planning to do extra changes, but later

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 7 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: it is originally done that way, reminder keeping as much original as possible

**Comment by gitcoindev** [comment]
Compare test coverage action failed with 0.1 % less coverage, I guess  we can neglect this due to amount of lines changed, but would be interesting to see which lines were affected.

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol (hunk 4 above)
Comment (molecula451) [reply packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:157]: some comments are misleading from the older commits but im recycling on a new PR

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol (hunk 5 above)
Comment (gitcoindev) [reply packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:349]: ok got it

File: packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol (hunk 4 above)
Comment (gitcoindev) [reply packages/contracts/src/dollar/libraries/LibDirectGovernanceFarmer.sol:157]: all right

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 6 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:200]: so basically changing the revert message

Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:200]: but the thing is also the duration can be 1 week only

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 7 above)
Comment (gitcoindev) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: I see originally minted amounts were 100 Dollars, 99 DAI, 98 USDC and 97 USDT and now 99 Dollars, 99 DAI, 98 USDC and 97 USDT. could you please double check if I see this right?

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 6 above)
Comment (gitcoindev) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:200]: > but the thing is also the duration can be 1 week only

require(
            durationWeeks >= 1 && durationWeeks <= 208,
            "duration weeks must be between 1 and 208"
        );

I meant that in this case we can test not only durationWeeks == 0 but also another boundary value durationWeeks == 209 , perhaps we can add this later in another PR as this one is just moving to facet.

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 7 above)
Comment (gitcoindev) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: Ok once we check and confirm this one I will approve from my side.

Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: both values shall be ok, the idea is to meet the assertion https://github.com/ubiquity/ubiquity-dollar/pull/841/files#diff-6140f6576af81654e444ba3b5bddd1fc61cfaf0a5670769f670bd2fb4ecb3527L331 that the user leaves no token in its wallet, because everything its deposited into the staking contract (as per original test), can you iterate it and  make the sense?

Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: i'm recycling a lot of bad commenting of new PR

Comment (gitcoindev) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:331]: @molecula451 all right, your explanations make sense. Since you are recycling bad commenting on new PR I will approve this one.

**Review by gitcoindev — APPROVED** [review]

**Comment by molecula451** [comment]
glad to have @gitcoindev on board lot of useful comments

**Comment by molecula451** [comment]
i think it'll make sense from a structured point of view to have all the facets ready before migrations? @rndquu

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 8]
```diff
@@ -569,20 +604,38 @@ contract DirectGovernanceFarmerTest is LocalTestHelper {
...
+        assertEq(dollar.balanceOf(userAddress), 100e18);
+        assertEq(token0.balanceOf(userAddress), 100e18);
+        assertEq(token1.balanceOf(userAddress), 100e18);
+        assertEq(token2.balanceOf(userAddress), 100e18);
     }
 
-    function testIsIdIncludedReturnTrueIfIdIsInTheList() public {
+    // END Multiple
+
+    /*function testIsIdIncludedReturnTrueIfIdIsInTheList() public {
```
Comment (rndquu) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: Why are these tests commented out?

File: packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol (hunk 1 above)
Comment (rndquu) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: The whole purpose of the `onERC1155Received()` method is to make sure that the contract supports (i.e. can receive) ERC1155 tokens so renaming `onERC1155Received()` to `onERC1155ReceivedDirectGovernance()`:
1. Violates [IERC1155Receiver](https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#IERC1155Receiver)
2. Breaks the contract in terms that it wouldn't support accepting ERC1155 tokens

The thing is that [DirectGovernanceFarmerFacet](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol) operates only with ERC20 tokens so I don't really understand why we even need to conform to `IERC1155Receiver`. I guess we can simply delete the [onERC1155ReceivedDirectGovernance](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol#L57) method.

**Review by rndquu — CHANGES_REQUESTED** [review]

**Comment by molecula451** [comment]
> The whole purpose of the onERC1155Received() method is to make sure that the contract supports (i.e. can receive) ERC1155 tokens so renaming onERC1155Received() to onERC1155ReceivedDirectGovernance():

> Violates [IERC1155Receiver](https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#IERC1155Receiver)
> Breaks the contract in terms that it wouldn't support accepting ERC1155 tokens
The thing is that [DirectGovernanceFarmerFacet](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol) operates only with ERC20 tokens so I don't really understand why we even need to conform to IERC1155Receiver. I guess we can simply delete the [onERC1155ReceivedDirectGovernance](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol#L57) method.

@rndquu 

Highlighting some things:

1. I think the main idea to have this method (as it is implemented in actual contract) is because and ERC1155  transfer happens under the hood when the user receives [LP Pool](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/DirectGovernanceFarmer.sol#L184) , at staking share transfer LibCreditNFTManager also [implements](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/libraries/LibCreditNftManager.sol#L242), the names cannot be the same in diamond facet
2. I couldn't just simple remove up until this PR happen due to keeping the originality of the contract
3.  Removing/purging this interface  is an idea coming on this PR

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 8 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: because this can be also set an assertion in depositSingle test when staking happens the user gets into the array ID list, maybe we want to have it alone or inside the [depositSingle ](https://github.com/ubiquity/ubiquity-dollar/pull/841/files#diff-6140f6576af81654e444ba3b5bddd1fc61cfaf0a5670769f670bd2fb4ecb3527R185)test maybe not?

File: packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol (hunk 1 above)
Comment (molecula451) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: @rndquu

_**Highlighting some things:**_

1. I think the main idea to have this method (as it is implemented in actual contract) is because and ERC1155 transfer happens under the hood when the user receives [LP Pool](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/DirectGovernanceFarmer.sol#L184) , at staking share transfer LibCreditNFTManager also [implements](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/libraries/LibCreditNftManager.sol#L242), the names cannot be the same in diamond facet
2. I couldn't just simple remove up until this PR happen due to keeping the originality of the contract
3. Removing/purging this interface is an idea coming on this PR

**Comment by molecula451** [comment]
@gitcoindev for your info, the tests that rndqnuu is highlighting (that are in discussion to keep it or not) are the reason the compare test coverage decreased

File: packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol (hunk 1 above)
Comment (rndquu) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: > I think the main idea to have this method (as it is implemented in actual contract) is because and ERC1155 transfer happens under the hood when the user receives [LP Pool](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/DirectGovernanceFarmer.sol#L184) , at staking share transfer LibCreditNFTManager also [implements](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/libraries/LibCreditNftManager.sol#L242), the names cannot be the same in diamond facet

I can't find where any ERC1155 transfers happen [here](https://github.com/ubiquity/ubiquity-dollar/blob/904396c1596128bb8f4b3bffde0ccb1f555a8451/packages/contracts/src/dollar/DirectGovernanceFarmer.sol#L184) under the hood. Could you show the related code?

> I couldn't just simple remove up until this PR happen due to keeping the originality of the contract

Again, it doesn't make sense in renaming `onERC1155Received()` to any other value because in this case the contract (the facet in our case) does not conform to [IERC1155Receiver](https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#IERC1155Receiver) hence `onERC1155ReceivedDirectGovernance()` **will never be called**.

I've just removed `onERC1155ReceivedDirectGovernance()` along with the `onERC1155Received()` [test](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol#L111) and all other tests are passing.

Comment (molecula451) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: Yeah it doesn't make sense either, and then im positive too in getting rid of this interface on this facet

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 8 above)
Comment (rndquu) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: It's not clear how [isIdIncluded()](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol#L161) is related to staking in terms of testing.

[isIdIncluded()](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol#L161) simply checks whether a uint value is in the list or not. This is basically what [these tests](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol#L638-L663) are checking. Why can't we use the same assertions but using the facet's [method](https://github.com/ubiquity/ubiquity-dollar/blob/c41a33daf263f569f9c7ef8a6ba3d7ff320ce7ed/packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol#L161) now?

Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: You get an ID `(typeof uint)` when you stake, the deposits functions all returns a [stakingShareId](https://github.com/ubiquity/ubiquity-dollar/pull/841/files#diff-4fe442d71f05b6caaf7a99183584f8e4f2b48eba36403391c2e8517a276a7551R95), the lists that you mention are the listed depositors so isIdIncluded checks if depositor is on the list, so it's a helper function, we could have an assertion in the deposit list test, (**which in fact tests that the helper **`isIdIncluded()`** is working**) or we could have the same test as the original contract (doing the same but like re-doing)

Comment (rndquu) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: Each test must check it's own method thus we separate concerns. If we assert `isIdIncluded()` in the `depositSingle()` method we violate separation of concerns. 

Basically we could assert all of the methods in a single test but we don't do it because this is hard to maintain hence we divide all tests to assert small pieces of logic and features.

Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: revamping this test to meet it's criteria

File: packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol (hunk 1 above)
Comment (molecula451) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: done in https://github.com/ubiquity/ubiquity-dollar/pull/841/commits/3554844f8f5dd7491b1d9f93d9b29c68cef4112f

Comment (molecula451) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: i have removed the interface but for your info here is a original comment (that i have not edited not update about the nft transfer) and it was implemented on the original contract
https://github.com/ubiquity/ubiquity-dollar/pull/841/files#diff-ec25ab7a74d12b4768ebfb4b5fe165eaff5dc3b3d90690074dc05a0fcc38b7a1L329

@rndquu re-thoughts are welcome

Comment (molecula451) [reply packages/contracts/src/dollar/facets/DirectGovernanceFarmerFacet.sol]: basically the bond is an ERC1155, so what it was implemented is the confirmation messaging (which this interface provide) that a ERC1155 was received

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 9]
```diff
@@ -614,30 +614,105 @@ contract DirectGovernanceFarmerFacetTest is DiamondTestSetup {
...
+        directGovernanceFarmerFacet.depositSingle(address(token0), 1, 1);
+        vm.stopPrank();
+
         // run assertions
         uint256[] memory list = new uint256[](1);
-        assertFalse(
-            directGovernanceFarmerHarness.exposed_isIdIncluded(list, 1)
-        );
-    }*/
+        assertFalse(directGovernanceFarmerFacet.isIdIncluded(list, 2));
```
Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:716]: @rndquu false, id (depositor) it's not on the list

Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:716]: @rndquu false, id (depositor) it's not on the list

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 10]
```diff
@@ -614,30 +614,105 @@ contract DirectGovernanceFarmerFacetTest is DiamondTestSetup {
...
+                "0x"
+            ),
+            ""
+        );
+
+        directGovernanceFarmerFacet.depositSingle(address(token0), 1, 1);
+        vm.stopPrank();
+
         // run assertions
         uint256[] memory list = new uint256[](1);
```
Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:715]: we mock an array list to memory to hold our uint values

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 11]
```diff
@@ -614,30 +614,105 @@ contract DirectGovernanceFarmerFacetTest is DiamondTestSetup {
...
+        directGovernanceFarmerFacet.depositSingle(address(token0), 1, 1);
+        vm.stopPrank();
         // run assertions
         uint256[] memory list = new uint256[](1);
         list[0] = 1;
-        assertTrue(directGovernanceFarmerHarness.exposed_isIdIncluded(list, 1));
+        assertTrue(directGovernanceFarmerFacet.isIdIncluded(list, 1));
     }
 
     function testIsIdIncludedReturnFalseIfIdIsNotInTheList() public {
```
Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:668]: we mimick the whole deposit stuff very important

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 9 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:716]: but since we checking for "2" then it's false because there is not 2 on our list as depositors, thus working up the helper function `isIdInclude()`

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 12]
```diff
@@ -614,30 +614,105 @@ contract DirectGovernanceFarmerFacetTest is DiamondTestSetup {
...
+                address(directGovernanceFarmerFacet),
+                userAddress,
+                1,
+                1,
+                "0x"
+            ),
+            ""
+        );
+
+        directGovernanceFarmerFacet.depositSingle(address(token0), 1, 1);
```
Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:660]: true, because our depositor (uint(1)) it's on our list!

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol [hunk 13]
```diff
@@ -614,30 +614,105 @@ contract DirectGovernanceFarmerFacetTest is DiamondTestSetup {
...
     // END Multiple
 
-    /*function testIsIdIncludedReturnTrueIfIdIsInTheList() public {
-        // deploy contract with exposed internal methods
-        DirectGovernanceFarmerHarness directGovernanceFarmerHarness = new DirectGovernanceFarmerHarness(
-                IUbiquityDollarManager(address(managerFacet)),
-                base3PoolAddress,
-                depositZapAddress
-            );
+    function testIsIdIncludedReturnTrueIfIdIsInTheList() public {
```
Comment (molecula451) [inline packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol:617]: mock a nice deposit

File: packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol (hunk 8 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: revamped at https://github.com/ubiquity/ubiquity-dollar/pull/841/commits/c29b403f0c8c2692c386b648545e0da11bd11b06

Comment (molecula451) [reply packages/contracts/test/diamond/facets/DirectGovernanceFarmerFacet.t.sol]: in terms of representation the depositors are forwarded as uint values thus getting arrayed on the list, this helper function a nice utility to facilitate the inclusion and returning a boolean to the value include

**Review by rndquu — APPROVED** [review]
